File: notes/javanotes/src/main/java/com/example/notesapp/service/NoteService.java
package com.example.notesapp.service;

import com.example.notesapp.model.Note;
import com.example.notesapp.repository.NoteRepository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.Optional;

public class NoteService {
    private final NoteRepository noteRepository;

    public NoteService(NoteRepository noteRepository) {
        this.noteRepository = java.util.Objects.requireNonNull(noteRepository, "noteRepository must not be null");
    }

    public Note createNote(String title, String content) {
        validateText(title, "title");
        validateText(content, "content");

        Note note = new Note(0L, title.trim(), content.trim(), LocalDateTime.now());
        return noteRepository.save(note);
    }

    public List<Note> getAllNotes() {
        return noteRepository.findAll();
    }

    public Optional<Note> findNoteById(long id) {
        return noteRepository.findById(id);
    }

    public boolean deleteNote(long id) {
        return noteRepository.deleteById(id);
    }

    public Note updateNote(long id, String title, String content) {
        validateText(title, "title");
        validateText(content, "content");

        Note existingNote = noteRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Note not found: " + id));
        Note updatedNote = new Note(existingNote.getId(), title.trim(), content.trim(), existingNote.getCreatedAt());
        return noteRepository.save(updatedNote);
    }

    public List<Note> searchNotes(String query) {
        String normalizedQuery = query == null ? "" : query.trim().toLowerCase(Locale.ROOT);
        if (normalizedQuery.isEmpty()) {
            return getAllNotes();
        }

        return noteRepository.findAll().stream()
                .filter(note -> note.getTitle().toLowerCase(Locale.ROOT).contains(normalizedQuery)
                        || note.getContent().toLowerCase(Locale.ROOT).contains(normalizedQuery))
                .toList();
    }

    private static void validateText(String value, String fieldName) {
        if (value == null || value.trim().isEmpty()) {
            throw new IllegalArgumentException(fieldName + " must not be blank");
        }
    }
}

File: notes/flutternotes/lib/services/note_service.dart
import '../models/note.dart';
import '../repositories/note_repository.dart';

class NoteService {
  final NoteRepository noteRepository;

  NoteService(this.noteRepository);

  Future<Note> createNote(String title, String content) async {
    _validateText(title, 'title');
    _validateText(content, 'content');

    final note = Note(
      id: 0,
      title: title.trim(),
      content: content.trim(),
      createdAt: DateTime.now(),
    );
    return noteRepository.save(note);
  }

  Future<List<Note>> getAllNotes() => noteRepository.findAll();

  Future<Note?> findNoteById(int id) => noteRepository.findById(id);

  Future<bool> deleteNote(int id) => noteRepository.deleteById(id);

  Future<Note> updateNote(int id, String title, String content) async {
    _validateText(title, 'title');
    _validateText(content, 'content');

    final existing = await noteRepository.findById(id);
    if (existing == null) {
      throw ArgumentError('Note not found: $id');
    }

    final updated = Note(
      id: existing.id,
      title: title.trim(),
      content: content.trim(),
      createdAt: existing.createdAt,
    );
    return noteRepository.save(updated);
  }

  Future<List<Note>> searchNotes(String query) async {
    final normalized = query.trim().toLowerCase();
    if (normalized.isEmpty) {
      return getAllNotes();
    }

    final all = await noteRepository.findAll();
    return all
        .where((note) =>
            note.title.toLowerCase().contains(normalized) ||
            note.content.toLowerCase().contains(normalized))
        .toList();
  }

  void _validateText(String value, String fieldName) {
    if (value.trim().isEmpty) {
      throw ArgumentError('$fieldName must not be blank');
    }
  }
}
